ibis/pandas/execution/timecontext.py

```diff
@@ -0,0 +1,111 @@
+""" Implementation of compute_time_context for time context related operations
+
+Time context of a node is computed at the beginning of execution phase.
+
+To use time context to load time series data:
+
+For operations like window, asof_join that adjust time context in execution,
+implement ``compute_time_context`` to pass different time contexts to child
+nodes.
+
+If ``pre_execute`` preloads any data, it should use timecontext to trim data
+to be in the time range.
+
+``execute_node`` of a leaf node can use timecontext to trim data, or to pass
+it as a filter in the database query.
+
+In some cases, data need to be trimmed in ``post_execute``.
+
+Note: In order to use the feature we implemented here, there should be a
+column of Timestamp type, and named as 'time' in TableExpr.
+See ``execute_database_table_client`` in ``generic.py``.
+And we assume timecontext is passed in as a tuple (begin, end) where begin and
+end are timestamp, or datetime string like "20100101".
+
+This is an optional feature. The result of executing an expression without time
+context is conceptually the same as executing an expression with (-inf, inf)
+time context.
+"""
+
+import numpy as np
+import pandas as pd
+
+import ibis.common.exceptions as com
+import ibis.expr.api as ir
+import ibis.expr.operations as ops
+from ibis.pandas.core import compute_time_context
+from ibis.pandas.execution import execute
+
+
+@compute_time_context.register(ops.AsOfJoin, list)
+def adjust_context_asof_join(op, computable_args, timecontext, **kwargs):
+    new_timecontexts = [timecontext for i in range(len(computable_args))]
+
+    if not timecontext:
+        return new_timecontexts
+
+    # right table should look back or forward
+    try:
+        begin, end = map(pd.to_datetime, timecontext)
```

Comment:
Do not do this here. timecontext should ready be converted.

Reply:
All checks and conversion are moved to the main execute step